compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1892,27 +1892,25 @@ object Types {
     /** The signature of the last known denotation, or if there is none, the
      *  signature of the symbol
      */
-    protected def computeSignature(implicit ctx: Context): Signature = {
-      val lastd = lastDenotation
-      if (lastd != null) lastd.signature
-      else symbol.asSeenFrom(prefix).signature
-    }
+    protected def computeSignature(implicit ctx: Context): Signature =
+      lastDenotation match
+        case null => symbol.asSeenFrom(prefix).initial.signature
+        case sd: SingleDenotation => sd.initial.signature
+        case d => d.signature
 
     /** The signature of the current denotation if it is known without forcing.
      *  Otherwise the signature of the current symbol if it is known without forcing.
      *  Otherwise NotAMethod.
      */
     private def currentSignature(implicit ctx: Context): Signature =
       if (ctx.runId == mySignatureRunId) mySignature
-      else {
-        val lastd = lastDenotation
-        if (lastd != null) lastd.signature
-        else {
+      else lastDenotation match
+        case null =>
           val sym = currentSymbol
-          if (sym.exists) sym.asSeenFrom(prefix).signature
+          if (sym.exists) sym.asSeenFrom(prefix).initial.signature
           else Signature.NotAMethod
-        }
-      }
+        case sd: SingleDenotation => sd.initial.signature
+        case d => d.signature
```

Comment:
why is the non-SingleDenotation case handled specially without calling initial ?

Reply:
We do not have an initial for non-singleton signatures

Reply:
OK, and it's fine because their signature is always `OverloadedSignature` right ? In that case maybe we should put a comment in the code or directly return `OverloadedSignature` to make that obvious